### PR TITLE
dump: add split big tidb query to several small queries

### PR DESCRIPTION
**Pull request by lichunzhu**
### What problem does this PR solve? 
Although we can make sure TiDB won't OOM through the help of TiDB tablesample. But if we didn't specify `--rows` to open tablesample, TiDB will still OOM.

### What is changed and how it works?
implement solution 2 in https://github.com/pingcap/dumpling/issues/232
Try to split the whole dump table SQL to several small queries to concatenate their results.

### Check List 

Tests 

 - Unit test
 - Integration test
 - Manual test:
 Test dumping 1.2TB data serially (with clustered index enabled), the data rows lines can match.
```log
[2021/03/28 22:31:00.483 +00:00] [INFO] [collector.go:212] ["backup Success summary: total backup ranges: 3, total success: 3, total failed: 0, total take(backup time): 7h12m34.659704485s, total take(real time): 7h12m34.659726417s, total size(MB): 1178105.36, avg speed(MB/s): 45.39, total rows: 2172298300"]
```
```
mysql> select count(*) from usertable;
+------------+
| count(*)   |
+------------+
| 2172298300 |
+------------+
1 row in set (11 min 59.50 sec)
```

Side effects

 - Increased code complexity

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation
 
### Release note

- add split big tidb query to several small queries to avoid TiDB OOM

### CODE REVIEW

**Comment by lichunzhu** [comment]
### Dump huge table test

Dump 1150.50GB clustered index data without -r argument
dumpling:v5.0.0-rc,mydumper will OOM.
dumpling:70a709a cost 7h12m34s, avg speed: 45.39 MB/s

### Comparison test result

1844.61MB data from a TiDB:v5.0.0-nightly, cluster indexed table. dump the data without -r argument
| binary | cost time | speed |
| ------ | --------- | ------  |
| dumpling:70a709a | 39.00s | 47.29 MB/s
| dumpling:v5.0.0-rc | 34.49s | 53.49 MB/s
| mydumper | 33.28s | 55.42 MB/s

dumpling:70a709a has some speed regression. Mainly because it opens and closes more mysql connections than other binaries.

File: v4/export/dump.go [hunk 1]
```diff
@@ -348,52 +410,74 @@ func (d *Dumper) sequentialDumpTable(tctx *tcontext.Context, conn *sql.Conn, met
...
+		}
+		tctx.L().Info("didn't build tidb concat sqls, will select all from table now",
+			zap.String("database", meta.DatabaseName()),
+			zap.String("table", meta.TableName()))
+	}
+	return d.dumpWholeTableDirectly(tctx, conn, meta, taskChan)
+}
+
+// concurrentDumpTable tries to split table into several chunks to dump
+// return (whether we should try to dump table serial, error)
```
Comment (lance6716) [inline v4/export/dump.go:438]: only one return value

File: v4/export/ir_impl.go [hunk 2]
```diff
@@ -51,6 +51,117 @@ func (iter *rowIter) HasNext() bool {
 	return iter.hasNext
 }
 
+// rowIter implements the SQLRowIter interface.
+// Note: To create a rowIter, please use `newRowIter()` instead of struct literal.
+type rowIterChunks struct {
```
Comment (lance6716) [inline v4/export/ir_impl.go:56]: maybe multiQueryChunkIter 🤔

File: v4/export/ir_impl.go [hunk 3]
```diff
@@ -215,3 +326,39 @@ func (m *metaData) MetaSQL() string {
 	}
 	return m.metaSQL
 }
+
+type tableDataChunks struct {
```
Comment (lance6716) [inline v4/export/ir_impl.go:330]: multiQueryChunk

File: v4/export/sql.go [hunk 4]
```diff
@@ -267,6 +268,36 @@ func SelectTiDBRowID(db *sql.Conn, database, table string) (bool, error) {
 	return true, nil
 }
 
+// GetSuitableRows gets suitable rows for each table
+func GetSuitableRows(db *sql.Conn, database, table string) uint64 {
+	const (
+		defaultRows = 200000
+		maxRows     = 1000000
+		sizePerFile = 128 * 1024 * 1024 // 128MB per file by default
```
Comment (lance6716) [inline v4/export/sql.go:276]: bytesPerFile

File: v4/export/sql.go [hunk 5]
```diff
@@ -267,6 +268,36 @@ func SelectTiDBRowID(db *sql.Conn, database, table string) (bool, error) {
 	return true, nil
 }
 
+// GetSuitableRows gets suitable rows for each table
+func GetSuitableRows(db *sql.Conn, database, table string) uint64 {
```
Comment (lance6716) [inline v4/export/sql.go:272]: context 🤔

File: v4/export/ir_impl.go [hunk 6]
```diff
@@ -215,3 +326,39 @@ func (m *metaData) MetaSQL() string {
...
+	tctx    *tcontext.Context
+	conn    *sql.Conn
+	queries []string
+	colLen  int
+	SQLRowIter
```
Comment (lance6716) [inline v4/export/ir_impl.go:335]: seems `SQLRowIter` this interface doesn't have a unit test

**Review by lance6716 — COMMENTED** [review]
rest lgtm

File: v4/export/ir_impl.go (hunk 2 above)
Comment (lance6716) [reply v4/export/ir_impl.go:56]: Queries

File: v4/export/sql.go (hunk 5 above)
Comment (lichunzhu) [reply v4/export/sql.go:272]: ?

File: v4/export/ir_impl.go [hunk 7]
```diff
@@ -51,6 +51,117 @@ func (iter *rowIter) HasNext() bool {
...
+
+func newRowIterChunks(tctx *tcontext.Context, conn *sql.Conn, queries []string, argLen int) *rowIterChunks {
+	r := &rowIterChunks{
+		tctx:    tctx,
+		conn:    conn,
+		queries: queries,
+		id:      0,
+		args:    make([]interface{}, argLen),
+	}
+	r.nextRows()
```
Comment (lance6716) [inline v4/export/ir_impl.go:75]: I was wandering why this iter needs an extra next. will review later

Comment (lichunzhu) [reply v4/export/ir_impl.go:75]: Because some sqls won't return result.

**Pull request merged**